calico_upgrade/pkg/migrate/felixconfig.go

```diff
@@ -0,0 +1,279 @@
+		isPtr := field.Type.Kind() == reflect.Ptr
+		fieldName := field.Name
+
+		switch {
+		case strings.HasPrefix(fieldName, "Failsafe"):
+			// Special-case the Failsafe ports - these require parsing and settings as a struct.
+			if configStrValue == "none" {
+				// Has no failsafe ports
+				vProtoPort := &[]apiv3.ProtoPort{}
+				fieldValue.Set(reflect.ValueOf(vProtoPort))
+				continue
+			}
+
+			vProtoPort, err := fc.parseProtoPort(configStrValue)
+			if err != nil {
+				data.ConversionErrors = append(data.ConversionErrors, ConversionError{})
+			}
+			fieldValue.Set(reflect.ValueOf(vProtoPort)) // pointer to proto port slice.
+			continue
+		case strings.HasPrefix(fieldName, "LogSeverity"):
+			// The log level fields need to have their value converted to the appropriate v3 value,
+			// but other than that are treated as normal string fields.
+			configStrValue = convertLogLevel(configStrValue)
+		}
+
+		// Set the field value based on the field type.
+		var kind reflect.Kind
+		if isPtr {
+			kind = field.Type.Elem().Kind()
+		} else {
+			kind = fieldValue.Kind()
+		}
+
+		switch kind {
+		case reflect.Uint32:
+			if value, err := strconv.ParseUint(configStrValue, 10, 32); err != nil {
+				continue
+			} else if isPtr {
+				vu := uint32(value)
+				fieldValue.Set(reflect.ValueOf(&vu))
+			} else {
+				fieldValue.SetUint(value)
+			}
+		case reflect.Int:
+			if value, err := strconv.ParseInt(configStrValue, 10, 64); err != nil {
+				continue
+			} else if isPtr {
+				vi := int(value)
+				fieldValue.Set(reflect.ValueOf(&vi))
+			} else {
+				fieldValue.SetInt(value)
+			}
+		case reflect.Bool:
+			if value, err := strconv.ParseBool(configStrValue); err != nil {
+			} else if isPtr {
+				fieldValue.Set(reflect.ValueOf(&value))
+			} else {
+				fieldValue.SetBool(value)
+			}
+		case reflect.String:
+			if isPtr {
+				fieldValue.Set(reflect.ValueOf(&configStrValue))
+			} else {
+				fieldValue.SetString(configStrValue)
+			}
+		default:
+			continue
+		}
+
+		// We must have set a field in the spec.
+		setField = true
+	}
+
+	if setField {
+		data.Resources = append(data.Resources, res)
+	}
+	return nil
+}
+
+func (fc *felixConfig) parseProtoPortFailed(msg string) error {
+	return errors.New(fmt.Sprintf("Failed to parse ProtoPort-%s", msg))
+}
+
+func (fc *felixConfig) parseProtoPort(raw string) (*[]apiv3.ProtoPort, error) {
+	var result []apiv3.ProtoPort
+	for _, portStr := range strings.Split(raw, ",") {
+		portStr = strings.Trim(portStr, " ")
+		if portStr == "" {
+			continue
+		}
+
+		parts := strings.Split(portStr, ":")
+		if len(parts) > 2 {
+			return nil, fc.parseProtoPortFailed("ports should be <protocol>:<number> or <number>")
+		}
+		protocolStr := "TCP"
+		if len(parts) > 1 {
+			protocolStr = strings.ToUpper(parts[0])
+			portStr = parts[1]
+		}
+		if protocolStr != "TCP" && protocolStr != "UDP" {
+			return nil, fc.parseProtoPortFailed("unknown protocol: " + protocolStr)
+		}
+
+		port, err := strconv.Atoi(portStr)
```

Comment:
good thing we didn't add named ports to v2.6.4. if we do add named ports to a 2.X branch, we'll have to remember to fix up the upgrade script ...